src/Repository/Geo/ZoneRepository.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace App\Repository\Geo;
+
+use App\Entity\Geo\Canton;
+use App\Entity\Geo\City;
+use App\Entity\Geo\CityCommunity;
+use App\Entity\Geo\ConsularDistrict;
+use App\Entity\Geo\Country;
+use App\Entity\Geo\Department;
+use App\Entity\Geo\District;
+use App\Entity\Geo\ForeignDistrict;
+use App\Entity\Geo\Region;
+use App\Entity\Geo\Zone;
+use App\Entity\Geo\ZoneableInterface;
+use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
+use Symfony\Bridge\Doctrine\RegistryInterface;
+
+final class ZoneRepository extends ServiceEntityRepository
+{
+    private const TYPES = [
+        Country::class => Zone::COUNTRY,
+        Region::class => Zone::REGION,
+        Department::class => Zone::DEPARTMENT,
+        District::class => Zone::DISTRICT,
+        Canton::class => Zone::CANTON,
+        CityCommunity::class => Zone::CITY_COMMUNITY,
+        City::class => Zone::CITY,
+        ForeignDistrict::class => Zone::FOREIGN_DISTRICT,
+        ConsularDistrict::class => Zone::CONSULAR_DISTRICT,
+    ];
+
+    public function __construct(RegistryInterface $registry)
+    {
+        parent::__construct($registry, Zone::class);
+    }
+
+    public function getTypeByZonable(ZoneableInterface $zoneable): string
+    {
+        $type = null;
+        foreach (self::TYPES as $class => $candidate) {
+            if ($zoneable instanceof $class) {
+                $type = $candidate;
+                break;
+            }
+        }
+
+        if (!$type) {
+            throw new \InvalidArgumentException(sprintf("Can't determine type for %s", \get_class($zoneable)));
+        }
+
+        return $type;
+    }
+
+    public function zoneableAsZone(ZoneableInterface $zoneable): Zone
+    {
+        $type = $this->getTypeByZonable($zoneable);
```

Comment:
what about adding a helper method to `ZoneableInterface`
```php
public function getType(): string;
```

and we can remove the method `getTypeByZonable` from the repository?

Reply:
added `public function getZoneType(): string;` to `ZoneableInterface`